components/server/src/main/java/com/hotels/styx/server/netty/codec/NettyToStyxRequestDecoder.java

```diff
@@ -78,29 +75,32 @@ private NettyToStyxRequestDecoder(Builder builder) {
     }
 
     @Override
-    protected void decode(ChannelHandlerContext ctx, HttpObject httpObject, List<Object> out) throws Exception {
-        if (httpObject.getDecoderResult().isFailure()) {
+    protected void decode(ChannelHandlerContext ctx, HttpObject httpObject, List<Object> out) {
+        if (httpObject.decoderResult().isFailure()) {
             String formattedHttpObject = httpMessageFormatter.formatNettyMessage(httpObject);
             throw new BadRequestException("Error while decoding request: " + formattedHttpObject,
-                    httpMessageFormatter.wrap(httpObject.getDecoderResult().cause()));
+                httpMessageFormatter.wrap(httpObject.decoderResult().cause()));
         }
 
         try {
             if (httpObject instanceof HttpRequest) {
-                this.producer = new FlowControllingHttpContentProducer(ctx, this.flowControlEnabled);
-                Observable<ByteBuf> contentObservable = Observable.create(contentSubscriber -> {
-                    contentSubscriber.setProducer(this.producer);
-                    this.producer.subscriptionStart(contentSubscriber);
-                });
+                contentBuffer = new FlowControllingHttpContentBuffer();
+                if (flowControlEnabled) {
+                    ctx.channel().config().setAutoRead(false);
+                }
+                Flux<ByteBuf> contentPublisher = Flux.create(contentBuffer::subscriptionStart)
+                    .doOnRequest(i -> ctx.channel().read())
+                    .doFinally(e -> ctx.channel().config().setAutoRead(true));
 
                 HttpRequest request = (HttpRequest) httpObject;
-                LiveHttpRequest styxRequest = toStyxRequest(request, contentObservable);
+                LiveHttpRequest styxRequest = toStyxRequest(request, contentPublisher);
                 out.add(styxRequest);
-            } else if (httpObject instanceof HttpContent && this.producer != null) {
-                this.producer.onNext(content(httpObject));
+            } else if (httpObject instanceof HttpContent) {
+                contentBuffer.addChunk(content(httpObject));
```

Comment:
`contentBuffer` can potentially be null?

Reply:
oh yes. I had to revert this part. thanks for catching